[WIP] Add basic heartbeat job tests

andrewvc opened the pull request:
**EDIT: The scope here is now much wider**

This PR is now about testing both HTTP and TCP, and introducing a new test schema DSL to do so. It's getting kinda big.

**Original below**

This PR adds some initial tests for the HTTP task. An open question is how comprehensive we want to be here and where we want to target these tests. It also breaks up one of the longer functions to make reasoning about it easier (and to make understanding the defer behavior more clear)

I think that thorough testing of each intermediate function is a bit overkill, so, I think the question is where to strike the balance. I would love your feedback @urso / @ruflin . 

My sense is that we definitely want some acceptance tests for each protocol that test that configuring them correctly does the right thing. 

We probably also want unit tests for most helper functions. For the actual IO however, I'm thinking we probably should only setup HTTP servers once in the tests simply to not waste time redundantly testing things. I don't think the spot I'm doing it in these examples is the best one. It probably makes more sense to set that same server up at the 'job level and test the `monitors.Job` object against it, including thorough tests of all expected fields.

Thoughts?

andrewvc commented on heartbeat/monitors/active/http/task_test.go:
```diff
@@ -84,27 +199,63 @@ func TestSplitHostnamePort(t *testing.T) {
+
+			if err != nil {
+				if test.expectedError == nil {
+					t.Error(err)
```
That's what blindly obeying your IDE gets you ;) Will fix

andrewvc commented on heartbeat/monitors/active/http/task.go:
```diff
@@ -218,41 +220,69 @@ func execPing(
-	defer resp.Body.Close()
 
 	err = validator(resp)
 	end = time.Now()
```
Good question! Any idea @urso ? I can't think of why we'd want to include the validator in the time. That said, I think the difference is negligible. Most validators should be dead simple.

ruflin commented:
Something still seems to break a system test: https://beats-ci.elastic.co/job/elastic+beats+pull-request+multijob-linux/5429/beat=heartbeat,label=ubuntu/testReport/junit/test_monitor/Test/test_http_1_404/ Let me know if you need some help to execute it locally to see all the logs etc.

andrewvc commented:
Thanks for the heads up. At the moment I'm reworking these tests into a
pseudo-framework. All of the end to end tests nest so to speak. An http
check should include a tcp check in the event.

I'll push up some larger changes EOD and take a look then

On Wed, Jul 11, 2018, 4:56 AM Nicolas Ruflin <notifications@github.com>
wrote:

> Something still seems to break a system test:
> https://beats-ci.elastic.co/job/elastic+beats+pull-request+multijob-linux/5429/beat=heartbeat,label=ubuntu/testReport/junit/test_monitor/Test/test_http_1_404/
> Let me know if you need some help to execute it locally to see all the logs
> etc.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/elastic/beats/pull/7551#issuecomment-404113365>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AAIBY1ajFjEQkFuFqAAosA-Tu34OD6Xuks5uFcvAgaJpZM4VItSE>
> .
>

andrewvc commented:
OK, so for the latest round of changes to the schema validation the code could definitely use a few more passes by myself before the internals are nice, but I'd love feedback on the externals at this point. How's this feel as a schema validation library from a user standpoint?

So, I got pretty far on the validation DSL that lets you match a nested map of validators to a MapStr. @urso and I agreed that this would be very useful for the sort of tests heartbeat needs. The following test suite tests the validation framework: https://github.com/andrewvc/beats/blob/add_hb_http_task_tests/heartbeat/skima/core_test.go

You can see how useful it is for heartbeat in the new basic tests I added for the http and tcp endpoints:
https://github.com/andrewvc/beats/blob/add_hb_http_task_tests/heartbeat/monitors/active/http/http_test.go#L66
https://github.com/andrewvc/beats/blob/add_hb_http_task_tests/heartbeat/monitors/active/tcp/tcp_test.go#L38

This obviously exceeds the original scope of this PR. I'll probably shut this one down soon and open a new one.

On @urso’s advice I modified the schema from my original procedural style to be more FP style and be oriented around function composition. I really like the result!

There’s some weirdness with typing ‘strict’ operations (validate that this map passes all checks and also doesn’t have additional keys). I implemented it two ways for now, you can either do it by composing `Strict(Schema(Map{…}))`, or, you can use `StrictMap{}` instead of `Map{}`, but due to the lack of union types that doesn’t work at the root level, so `Schema(StrictMap{...})` doesn’t work, but submaps can be strict. There are some ways around it (just use interface{})? but probably worth a discussion. I kind of like keeping both options since internally `StrictMap` is implemented with `Strict(Schema)`.  It's also nice because when you use shared schemas and compose them, you might not know whether you want strict validation at schema creation time or not. In other words, the flexibility is nice.

andrewvc commented:
This thread is too long. Moved to: https://github.com/elastic/beats/pull/7587
